Task 3/Task 3.2/Task 3.2/Task 3.2/DynamicArray.cs

```diff
@@ -0,0 +1,250 @@
+                        {
+                            _myArray[index] = item;
+                        }
+                        else
+                        {
+                            _myArray[i] = _myArray[i - 1];
+                        }
+                    }
+                    length++;
+                    return true;
+                }
+                else
+                {
+                    T[] array = new T[capacity * 2];
+                    for (int i = 0; i <= length; i++)
+                    {
+                        if (i < index)
+                            array[i] = _myArray[i];
+                        else if (i == index)
+                            array[i] = item;
+                        else
+                            array[i] = _myArray[i - 1];
+                    }
+                    _myArray = array;
+                    capacity = array.Length;
+                    length++;
+                    return true;
+                }
+            }
+        }
+
+        public void SetCapacity(int value)
+        {
+            if (value >= length)
+            {
+                T[] array = new T[value];
+                for (int i = 0; i < length; i++)
+                {
+                    array[i] = _myArray[i];
+                }
+                _myArray = array;
+                capacity = value;
+            }
+            else
+            {
+                T[] array = new T[value];
+                for (int i = 0; i < array.Length; i++)
+                {
+                    array[i] = _myArray[i];
+                }
+                _myArray = array;
+                capacity = value;
+                length = array.Length;
+            }
+        }
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            for (int i = 0; i < length; i++)
+            {
+                yield return _myArray[i];
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return _myArray.GetEnumerator();
+        }
+
+        public object Clone()
+        {
+            DynamicArray<T> newArray = new DynamicArray<T>(capacity) { };
+            for (int i = 0; i < capacity; i++)
+            {
+                newArray._myArray[i] = _myArray[i];
+            }
+            return newArray;
+        }
+
+        public T[] ToArray()
+        {
+            T[] newArray = new T[length];
+
+            for (int i = 0; i < length; i++)
+            {
+                newArray[i] = _myArray[i];
+            }
+
+            return newArray;
+        }
+
+        public T this[int index]
+        {
+            get
+            {
+                if (index >= length)
+                    throw new ArgumentOutOfRangeException();
+                else if (index < 0)
+                {
+                    if (Math.Abs(index) > length)
+                    {
+                        throw new ArgumentOutOfRangeException();
+                    }
```

Comment:
You could throw only once by using Abs. (otherwise in `else if` `index < -length` would be enough)